phoenix-core/src/main/java/org/apache/phoenix/iterate/ClientHashAggregatingResultIterator.java

```diff
@@ -0,0 +1,155 @@
+    private final List<Expression> groupByExpressions;
+    private HashMap<ImmutableBytesWritable, Aggregator[]> hash;
+    private List<ImmutableBytesWritable> keyList;
+    private Iterator<ImmutableBytesWritable> keyIterator;
+
+    public ClientHashAggregatingResultIterator(ResultIterator resultIterator, Aggregators aggregators, List<Expression> groupByExpressions) {
+        if (resultIterator == null) throw new NullPointerException();
+        if (aggregators == null) throw new NullPointerException();
+        if (groupByExpressions == null) throw new NullPointerException();
+        this.resultIterator = resultIterator;
+        this.aggregators = aggregators;
+        this.groupByExpressions = groupByExpressions;
+    }
+
+    @Override
+    public Tuple next() throws SQLException {
+        if (keyIterator == null) {
+            populateHash();
+            sortKeys();
+            keyIterator = keyList.iterator();
+        }
+
+        if (!keyIterator.hasNext()) {
+            return null;
+        }
+
+        ImmutableBytesWritable key = keyIterator.next();
+        Aggregator[] rowAggregators = hash.get(key);
+        byte[] value = aggregators.toBytes(rowAggregators);
+        Tuple tuple = wrapKeyValueAsResult(KeyValueUtil.newKeyValue(key, SINGLE_COLUMN_FAMILY, SINGLE_COLUMN, AGG_TIMESTAMP, value, 0, value.length));
+        return tuple;
+    }
+
+    @Override
+    public void close() throws SQLException {
+        keyIterator = null;
+        keyList = null;
+        hash = null;
+        resultIterator.close();
+    }
+    
+    @Override
+    public Aggregator[] aggregate(Tuple result) {
+        Aggregator[] rowAggregators = aggregators.getAggregators();
+        aggregators.reset(rowAggregators);
+        aggregators.aggregate(rowAggregators, result);
+        return rowAggregators;
+    }
+
+    @Override
+    public void explain(List<String> planSteps) {
+        resultIterator.explain(planSteps);
+    }
+
+    @Override
+    public String toString() {
+        return "ClientHashAggregatingResultIterator [resultIterator=" 
+            + resultIterator + ", aggregators=" + aggregators + ", groupByExpressions="
+            + groupByExpressions + "]";
+    }
+
+    // Copied from ClientGroupedAggregatingResultIterator
+    protected ImmutableBytesWritable getGroupingKey(Tuple tuple, ImmutableBytesWritable ptr) throws SQLException {
+        try {
+            ImmutableBytesWritable key = TupleUtil.getConcatenatedValue(tuple, groupByExpressions);
+            ptr.set(key.get(), key.getOffset(), key.getLength());
+            return ptr;
+        } catch (IOException e) {
+            throw new SQLException(e);
+        }
+    }
+
+    // Copied from ClientGroupedAggregatingResultIterator
+    protected Tuple wrapKeyValueAsResult(KeyValue keyValue) {
+        return new MultiKeyValueTuple(Collections.<Cell> singletonList(keyValue));
+    }
+
+    private void populateHash() throws SQLException {
```

Comment:
For a large result set of join, this code snippet has a risk lead to OOM on client. Maybe add limitation is  more graceful

Reply:
Hi @solzy , thanks for the review. My understanding of the pre-existing code is that classes like OrderedResultIterator do not handle OOM, and they do not provide a limitation. I don't want to introduce a mechanism that is specific to this feature. Is there any place where the client handles OOM, which can be used as a reference?

Reply:
@geraldss Sorry for late reply.  i did not find some handle for OOM on the client, which will be affects stability  when uses as QueryServer. anyway, we can improve these features in the future. Thanks for your response , good job.

Reply:
@geraldss - memory management is tracked by the GlobalMemoryManager. Operations that potentially use memory allocate (and eventually free) a set of MemoryChunk instances. You can see an example of this in GroupedAggregateRegionObserver (the runtime code for aggregation). If the memory used goes over a threshold (phoenix.query.maxGlobalMemoryPercentage and phoenix.query.maxTenantMemoryPercentage as the allowed percentage of Java heap across all queries that is allowed to be used), then the query will fail. Most typically, this mechanism is used on the server side as we don't typically use a lot of memory on the client-side (as we're mostly doing merge joins). One example where we use this on the client side is for our broadcast join implementation (see HashCacheClient) to track memory held onto for Hash Join caches.

Classes you may want to look at (or perhaps you already have?): OrderedResultIterator and MappedByteBufferSortedQueue. Above a certain configurable threshold (phoenix.query.spoolThresholdBytes defaults to 20MB), we output results into memory mapped files. Have you tried decreasing that threshold?
 
Couple of JIRAs you may want to take a look at: PHOENIX-2405 (unclear if this is still an issue) and PHOENIX-3289. Are you running into issues with too many memory mapped files?

Reply:
Hi @JamesRTaylor, thanks for the review and feedback. I'm hoping to get a minimum implementation merged into master, and it can then improve over time. I see that HashCacheClient throws an exception after the memory threshold is reached. Would it suffice if I do the same thing?

The other examples, e.g. MappedByteBufferSortedQueue, involve linear memory.

Reply:
@JamesRTaylor @solzy I have added a memory limitation based on SPOOL_THRESHOLD_BYTES. It throws an exception similar to HashCacheClient. Please review. Thanks.